market-generator/src/test/java/com/market/banica/generator/service/MarketServiceImplTest.java

```diff
@@ -114,4 +141,121 @@ void request_ReturnSuperRequest() {
+        verify(marketState, times(1)).removeItemFromState(GOOD_BANICA, AMOUNT, PRICE_1);
+    }
+
+    @Test
+    public void checkAvailabilityIncreasesPreviousRecordQuantityInPendingOrders() throws ProductNotAvailableException {
+
+        ProductBuySellRequest availabilityRequest = populateBuySellRequest(GOOD_BANICA, AMOUNT, PRICE_1);
+
+        MarketTick marketTick = new MarketTick(GOOD_BANICA, AMOUNT, PRICE_1, TIMESTAMP);
+
+        when(marketState.removeItemFromState(GOOD_BANICA, AMOUNT, PRICE_1)).thenReturn(marketTick);
+
+        marketService.checkAvailability(availabilityRequest, calculatorAvailability);
+        marketService.checkAvailability(availabilityRequest, calculatorAvailability);
+
+        MarketTick marketTickWithIncreasedQuantity = pendingOrders.get(GOOD_BANICA).get(PRICE_1);
+
+        assertEquals(1, pendingOrders.size());
+        assertEquals(2, marketTickWithIncreasedQuantity.getQuantity());
+        verify(marketState, times(2)).removeItemFromState(GOOD_BANICA, AMOUNT, PRICE_1);
+    }
+
+    @Test
+    public void checkAvailabilityAddsSecondRecordInPendingOrders() throws ProductNotAvailableException {
+
+        ProductBuySellRequest availabilityRequest = populateBuySellRequest(GOOD_BANICA, AMOUNT, PRICE_1);
+        ProductBuySellRequest secondAvailabilityRequest = populateBuySellRequest(GOOD_BANICA, AMOUNT, PRICE_2);
+
+        MarketTick marketTick = new MarketTick(GOOD_BANICA, AMOUNT, PRICE_1, TIMESTAMP);
+        MarketTick secondMarketTick = new MarketTick(GOOD_BANICA, AMOUNT, PRICE_2, TIMESTAMP);
+
+        when(marketState.removeItemFromState(GOOD_BANICA, AMOUNT, PRICE_1)).thenReturn(marketTick);
+        when(marketState.removeItemFromState(GOOD_BANICA, AMOUNT, PRICE_2)).thenReturn(secondMarketTick);
+
+        marketService.checkAvailability(availabilityRequest, calculatorAvailability);
+        marketService.checkAvailability(secondAvailabilityRequest, calculatorAvailability);
+
+        assertEquals(2, pendingOrders.get(GOOD_BANICA).size());
+        verify(marketState, times(1)).removeItemFromState(GOOD_BANICA, AMOUNT, PRICE_1);
+        verify(marketState, times(1)).removeItemFromState(GOOD_BANICA, AMOUNT, PRICE_2);
+    }
+
+    @Test
+    public void returnPendingProductClearsPendingOrders() {
+
+        MarketTick marketTick = new MarketTick(GOOD_BANICA, 10, PRICE_1, TIMESTAMP);
+
+        pendingOrders.put(GOOD_BANICA, new TreeMap<>());
+        pendingOrders.get(GOOD_BANICA).put(PRICE_1, marketTick);
+
+        ProductBuySellRequest request = populateBuySellRequest(GOOD_BANICA, 10, PRICE_1);
+
+        marketService.returnPendingProduct(request, calculatorReturnProduct);
+
+        assertEquals(0, pendingOrders.size());
+    }
+
+    @Test
+    public void returnPendingProductRemovesMarketTickFromPendingOrders() {
+
+        long amount = 10;
+
+        MarketTick marketTick = new MarketTick(GOOD_BANICA, amount, PRICE_1, TIMESTAMP);
```

Comment:
Market ticks should not have same timestamp and adding an additional compare with price is not a quickfix.

Reply:
Fixed